### PR TITLE
Support switching on / off UDT together with in-Memtable-only feature

### PR DESCRIPTION
**Pull request by jowlyzhang**
Add support to allow enabling / disabling user-defined timestamps feature for an existing column family in combination with the in-Memtable only feature.

To do this, this PR includes:
1) Log the `persist_user_defined_timestamps` option per column family in Manifest to facilitate detecting an attempt to enable / disable UDT. This entry is enforced to be logged in the same VersionEdit as the user comparator name entry.

2) User-defined timestamps related options are validated when re-opening a column family, including user comparator name and the `persist_user_defined_timestamps` flag. These type of settings and settings change are considered valid:
     a) no user comparator change and no effective `persist_user_defined_timestamp` flag change.
     b) switch user comparator to enable UDT provided the immediately effective `persist_user_defined_timestamps` flag 
         is false.
     c) switch user comparator to disable UDT provided that the before-change `persist_user_defined_timestamps` is 
         already false.
3) when an attempt to enable UDT is detected, we mark all its existing SST files as "having no UDT" by marking its `FileMetaData.user_defined_timestamps_persisted` flag to false and handle their file boundaries `FileMetaData.smallest`, `FileMetaData.largest` by padding a min timestamp.

4) while enabling / disabling UDT feature, timestamp size inconsistency in existing WAL logs are handled to make it compatible with the running user comparator.

Test plan:
```
make all check
./db_with_timestamp_basic_test --gtest-filter="*EnableDisableUDT*"
./db_wal_test --gtest_filter="*EnableDisableUDT*"
```

### CODE REVIEW

**Review by ltamasi — APPROVED** [review]
Looks great, thanks @jowlyzhang !

File: db/db_impl/db_impl_open.cc [hunk 1]
```diff
@@ -1206,35 +1206,37 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
 
       // We create a new batch and initialize with a valid prot_info_ to store
       // the data checksums
-      WriteBatch batch;
+      std::unique_ptr<WriteBatch> batch(new WriteBatch());
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/db_impl/db_impl_open.cc]: This is a good idea! Thank you for the suggestion. It helps to save an extra argument for the util function too.

File: db/db_wal_test.cc [hunk 2]
```diff
@@ -363,56 +358,55 @@ class DBWALTestWithTimestamp
...
     ASSERT_OK(Put(1, "foo", ts2, "v3"));
 
-    ASSERT_OK(ReopenColumnFamiliesWithTs({"pikachu"}, ts_options,
-                                         avoid_flush_during_recovery));
+    ASSERT_OK(ReopenColumnFamiliesWithDefaultCF({"pikachu"}, ts_options,
+                                                avoid_flush_during_recovery));
     ASSERT_EQ(GetNumberOfSstFilesForColumnFamily(db_, "pikachu"), 0U);
-    std::string ts3 = Timestamp(3, 0);
+    std::string ts3;
+    PutFixed64(&ts3, 3);
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/db_wal_test.cc:415]: Thank you for helping thoroughly review the test cases. You are right that the passed timestamps are confusing, I got lost myself too when double checking it. I added some comments to the test case to help navigate it.

File: util/udt_util.cc [hunk 3]
```diff
@@ -100,6 +100,45 @@ Status CheckWriteBatchTimestampSizeConsistency(
...
+    return ToggleUDT::kEnableUDT;
+  }
+  if (new_name_size + kSuffixSize == old_name_size &&
+      IsPrefix(old_comparator_name.data(), new_comparator_name) &&
+      IsPrefix(old_comparator_name.data() + new_name_size, kUDTSuffix)) {
+    assert(ts_sz == 0);
+    return ToggleUDT::kDisableUDT;
+  }
+  return ToggleUDT::kInvalidChange;
+}
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply util/udt_util.cc]: It looks much more idiomatic with this change. Thank you for the suggestion!

**Pull request closed without merging**